### DISCUSSION TITLE
`null`免除演算子

### DISCUSSION THREAD

**Discussion by taiseiue** [WSOFT-Project/alicescript#20] [Ideas]
いい感じにnull許容型の値を取り出せる演算子が欲しい

```cs
number? num = 123;

number x = num.Value;

x = num!;
```

上のコードで、`x`に値が設定されている(=x.HasValueがtrue)な時、`x.Value`の値を返して、値がないときはnull参照エラーにするのが良さそう。

ユーザー入力がらみで`null`許容な値が出てくることが割とあって、頻繁に書く割には文法として楽じゃないのが良くなさげ。
例えば以下のコードは割と良くある。

```cs
number? rawInput = input() as number;

if(rawInput.HasValue)
// ユーザーがちゃんと数値を入力した場合
```

**Comment by rokuosan** [comment]
非nullアサーションのような方が良さそう
https://kotlinlang.org/docs/null-safety.html#not-null-assertion-operator

Kotlin の Null-Safety を実現するための仕組みはこのページにあるので一回眺めるだけでも良さそうかも

**Reply by rokuosan** [reply]
意図が違ったら申し訳ない

**Comment by taiseiue** [comment]
ここで考えるべきことは、`null`許容なコンテキストと非許容なコンテキストを明確に分けるためにnull許容から非nullへの暗黙的な変換を禁止してるから、そんなにお手軽に`null`免除アサーションできていいのかという
